### PR TITLE
Fix/consensus order

### PR DESCRIPTION
**Pull request by kamilsa**
### Description of the Change

Currently yac consensus's function currentLeader does not have check on whether order vector is empty or not. This PR adds this check by means of optional.

### Benefits

Error handling on when order vector is empty.

### Possible Drawbacks 

Too many places in code, needed to be updated because of that change

### CODE REVIEW

File: irohad/consensus/yac/cluster_order.hpp [hunk 1]
```diff
@@ -36,7 +36,7 @@ namespace iroha {
         /**
          * Provide current leader peer
          */
-        model::Peer currentLeader();
+        nonstd::optional<model::Peer> currentLeader();
```
Comment (l4l) [inline irohad/consensus/yac/cluster_order.hpp]: 1. Why not boost?
2. Header seems missing

File: irohad/consensus/yac/impl/yac.cpp [hunk 2]
```diff
@@ -128,12 +128,15 @@ namespace iroha {
...
-          timer_->invokeAfterDelay(delay_,
-                                   [this, vote] { this->votingStep(vote); });
-        }
+
+        cluster_order_.currentLeader() | [&](auto leader){
+          network_->send_vote(leader, vote);
+          cluster_order_.switchToNext();
+          if (cluster_order_.hasNext()) {
+            timer_->invokeAfterDelay(delay_,
+                                     [this, vote] { this->votingStep(vote); });
```
Comment (l4l) [inline irohad/consensus/yac/impl/yac.cpp]: capture `vote` by-ref

File: irohad/consensus/yac/cluster_order.hpp (hunk 1 above)
Comment (kamilsa) [reply irohad/consensus/yac/cluster_order.hpp]: Which header?

Comment (l4l) [reply irohad/consensus/yac/cluster_order.hpp]: for optional

File: irohad/consensus/yac/cluster_order.hpp [hunk 3]
```diff
@@ -49,13 +55,20 @@ namespace iroha {
...
-        auto getNumberOfPeers() { return order_.size(); }
+        auto getNumberOfPeers() {
+          return order_.size();
+        }
 
         virtual ~ClusterOrdering() = default;
 
        private:
+        explicit ClusterOrdering(std::vector<model::Peer> order);
+        ClusterOrdering() {}
```
Comment (muratovv) [inline irohad/consensus/yac/cluster_order.hpp]: what is the reason for presence of empty ctor here?

File: irohad/consensus/yac/impl/cluster_order.cpp [hunk 4]
```diff
@@ -15,19 +15,26 @@
...
-#include <utility>
 #include "consensus/yac/cluster_order.hpp"
 
 namespace iroha {
   namespace consensus {
     namespace yac {
 
+      nonstd::optional<ClusterOrdering> ClusterOrdering::create(
+          std::vector<model::Peer> order) {
+        if (order.empty()) {
```
Comment (muratovv) [inline irohad/consensus/yac/impl/cluster_order.cpp]: minor issue: `if` statement may be simplified with the ternary operator.

File: test/module/irohad/consensus/yac/yac_mocks.hpp [hunk 5]
```diff
@@ -139,9 +145,9 @@ namespace iroha {
 
         MockYacPeerOrderer() = default;
 
-        MockYacPeerOrderer(const MockYacPeerOrderer &rhs) {};
+        MockYacPeerOrderer(const MockYacPeerOrderer &rhs){};
```
Comment (muratovv) [inline test/module/irohad/consensus/yac/yac_mocks.hpp]: Is it changes happen with clang-format formatting? If no, check the whole pr and fix it.

File: irohad/consensus/yac/cluster_order.hpp (hunk 3 above)
Comment (kamilsa) [reply irohad/consensus/yac/cluster_order.hpp]: To make sure that creation of this object is only possible with create method

File: test/module/irohad/consensus/yac/yac_mocks.hpp (hunk 5 above)
Comment (kamilsa) [reply test/module/irohad/consensus/yac/yac_mocks.hpp]: Yes, because of clang-formatting

File: irohad/consensus/yac/cluster_order.hpp (hunk 3 above)
Comment (Warchant) [reply irohad/consensus/yac/cluster_order.hpp]: public:
ClusterOrdering() = delete;

Comment (Warchant) [reply irohad/consensus/yac/cluster_order.hpp]: https://stackoverflow.com/questions/13654927/why-explicitly-delete-the-constructor

**Comment by l4l** [comment]
Some test also will be nice to add here

File: irohad/consensus/yac/cluster_order.hpp [hunk 6]
```diff
@@ -31,7 +31,13 @@ namespace iroha {
...
        public:
 
-        explicit ClusterOrdering(std::vector<model::Peer> order);
+        /**
+         * Creates cluster ordering from the vector of peers
+         * @param order vector of peers
+         * @return false if vector is empty, true otherwise
+         */
+        static nonstd::optional<ClusterOrdering> create(
+            std::vector<model::Peer> order);
```
Comment (l4l) [inline irohad/consensus/yac/cluster_order.hpp]: Can be const-ref

**Pull request merged**